controllers/virtualmachine/volumebatch/volumebatch_controller.go

```diff
@@ -0,0 +1,663 @@
+	pvcKey := client.ObjectKey{
+		Namespace: ctx.VM.Namespace,
+		Name:      pvcName,
+	}
+
+	if err := r.Get(ctx, pvcKey, pvc); err != nil {
+		return err
+	}
+
+	if v, ok := pvc.Spec.Resources.Requests[corev1.ResourceStorage]; ok {
+		// Use the request if it exists.
+		status.Requested = &v
+	}
+
+	if v, ok := pvc.Spec.Resources.Limits[corev1.ResourceStorage]; ok {
+		// Use the limit if it exists.
+		status.Limit = &v
+	} else {
+		// Otherwise use the requested capacity.
+		status.Limit = status.Requested
+	}
+
+	return nil
+}
+
+// ReconcileDelete handles cleanup when a VirtualMachine is being deleted.
+// This method ensures proper cleanup of batch attachments and related resources.
+func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VolumeContext) error {
+	// TODO: Implement VM deletion cleanup logic
+	// This method should:
+	// - Clean up CnsNodeVmBatchAttachment resources
+	// - Handle any pending volume operations
+	// - Ensure proper resource cleanup on VM deletion
+
+	return nil
+}
+
+// reconcileInstanceStoragePVCs handles instance storage PVC lifecycle management.
+// This provides feature parity with the v1 controller's instance storage support.
+func (r *Reconciler) reconcileInstanceStoragePVCs(_ *pkgctx.VolumeContext) (bool, error) {
+	// TODO: Implement instance storage PVC reconciliation
+	// This method should:
+	// - Create missing instance storage PVCs
+	// - Handle PVC binding and placement
+	// - Manage instance storage annotations
+	// - Handle placement failures and cleanup
+
+	return true, nil
+}
+
+// handlePVCWithWFFC handles PVCs with WaitForFirstConsumer binding mode.
+// This ensures proper node selection for storage classes requiring it.
+func (r *Reconciler) handlePVCWithWFFC(_ *pkgctx.VolumeContext, _ vmopv1.VirtualMachineVolume) error {
+	// TODO: Implement WFFC (WaitForFirstConsumer) handling
+	// This method should:
+	// - Check storage class binding mode
+	// - Set selected node annotations for unbound PVCs
+	// - Handle node selection for proper PVC binding
+	// - Validate node availability and zone constraints
+
+	return nil
+}
+
+// cleanupOrphanedVolumes removes volumes that are no longer in the VM spec.
+// This handles the cleanup of volumes that have been removed from the desired state.
+func (r *Reconciler) cleanupOrphanedVolumes(_ *pkgctx.VolumeContext, _ []vmopv1.VirtualMachineVolume) error {
+	// TODO: Implement orphaned volume cleanup
+	// This method should:
+	// - Identify volumes in batch attachment but not in VM spec
+	// - Remove orphaned volumes from batch attachment
+	// - Preserve status information during cleanup
+	// - Handle cleanup errors gracefully
+
+	return nil
+}
+
+// processVolumesWithErrorHandling processes volumes with robust error handling.
+// This ensures partial failures don't prevent other volumes from being processed.
+func (r *Reconciler) processVolumesWithErrorHandling(_ *pkgctx.VolumeContext, _ []vmopv1.VirtualMachineVolume) error {
```

Comment:
Do we still need to make sure only one `pending` attachement at a time as previous logic? `onlyAllowOnePendingAttachment`

Reply:
I will leave a TODO for this.   Even that logic was a hack at best so I would want to explore what breaks if we get rid of that.